src/dune_rules/coq/coq_lib.ml

```diff
@@ -56,51 +70,155 @@ include struct
           (* whether this theory uses the stdlib, eventually set to false for all libs *)
     ; src_root : Path.Build.t
     ; obj_root : Path.Build.t
-    ; theories : (Loc.t * t) list Resolve.t
+    ; theories : (Loc.t * R.t) list Resolve.t
     ; libraries : (Loc.t * Lib.t) list Resolve.t
-    ; theories_closure : t list Resolve.t Lazy.t
+    ; theories_closure : R.t list Resolve.t Lazy.t
     ; package : Package.t option
     }
+
+  val to_dyn : t -> Dyn.t
+
+  val src_root : t -> Path.Build.t
+
+  val obj_root : t -> Path.Build.t
+
+  val implicit : t -> bool
+end = struct
+  type t =
+    { loc : Loc.t
+    ; boot_id : Id.t option
+          (** boot library that was selected to build this theory *)
+    ; id : Id.t
+    ; implicit : bool (* Only useful for the stdlib *)
+    ; use_stdlib : bool
+          (* whether this theory uses the stdlib, eventually set to false for all libs *)
+    ; src_root : Path.Build.t
+    ; obj_root : Path.Build.t
+    ; theories : (Loc.t * R.t) list Resolve.t
+    ; libraries : (Loc.t * Lib.t) list Resolve.t
+    ; theories_closure : R.t list Resolve.t Lazy.t
+    ; package : Package.t option
+    }
+
+  let to_dyn
+      { loc
+      ; boot_id
+      ; id
+      ; implicit
+      ; use_stdlib
+      ; src_root
+      ; obj_root
+      ; theories
+      ; libraries
+      ; theories_closure
+      ; package
+      } =
+    Dyn.(
+      record
+        [ ("loc", Loc.to_dyn loc)
+        ; ("boot_id", Dyn.option Id.to_dyn boot_id)
+        ; ("id", Id.to_dyn id)
+        ; ("implicit", Bool.to_dyn implicit)
+        ; ("use_stdlib", Bool.to_dyn use_stdlib)
+        ; ("src_root", Path.Build.to_dyn src_root)
+        ; ("obj_root", Path.Build.to_dyn obj_root)
+        ; ( "theories"
+          , Resolve.to_dyn (Dyn.list (Dyn.pair Loc.to_dyn R.to_dyn)) theories )
+        ; ( "libraries"
+          , Resolve.to_dyn (Dyn.list (Dyn.pair Loc.to_dyn Lib.to_dyn)) libraries
+          )
+        ; ( "theories_closure"
+          , Resolve.to_dyn (Dyn.list R.to_dyn) (Lazy.force theories_closure) )
+        ; ("package", Dyn.option Package.to_dyn package)
+        ])
+
+  let src_root t = t.src_root
+
+  let obj_root t = t.obj_root
+
+  let implicit t = t.implicit
 end
 
-let name l = l.id.name
+and Legacy : sig
+  type t =
+    { boot_id : Id.t option
```

Comment:
I guess that legacy libraries are always built with the coqlib boot library, so I'm not sure this field makes sense, right?

Actually there are quite a few corner cases here with the coqlib, in the sense that even having a boot library in scope doesn't mean we should reject what we have in user contrib actually, as for example there could be some useful stuff like HoTT globally installed, so that's kind of tricky.

So yeah, maybe the field is ok even if it is going to be always set to $coqlib/theories.